README.md
# katex\_flutter

Render scientific LaTeX equations using the KaTeX library.

- **Mathematics / Maths Equations** (Algebra, Calculus, Geometry, Geometry etc...)
- **Physics Equations**
- **Signal Processing Equations**
- **Chemistry Equations**
- **Statistics / Stats Equations**
- **Inherit text style from parent widgets**

*Note: The bad pub.dev score is due to a known error in the analysis server and not our fault.*

LaTeX rendering is done using **[CaTeX](https://pub.dev/packages/catex)** and **[KaTeX](https://github.com/KaTeX/KaTeX)**.

**katex_flutter** is working on Android, iOS, the Web and the Desktop platform.

We are now using native Flutter widgets to render the LaTeX code. This significantly increates the rendering performance
compared to previous versions. For unsupported LaTeX commands, we offer a fallback to the old JavaScript library.

## API

**Note: we recently did significant API changes. Please update any prior code.**
```dart
KaTeX(
  @required laTeXCode,         // A Text() containing the LaTeX code to be rendered
  delimiter = r'$',            // The delimiter to be used for inline LaTeX
  displayDelimiter = r'$$',    // The delimiter to be used for Display (centered, "important") LaTeX
  );
```

## What's the difference to CaTeX library?

Basically, **katex_flutter** is using CaTeX. But as CaTeX is still under development, we
offer a hybrid solution: As long as your LaTeX input is supported by CaTeX, you can profit
from CaTeX' performance but in case CaTeX does not fully support your input, you may fall
back on a slower but more complete LaTeX implementation using JavaScript and the KaTeX library.

Another difference is Text support: CaTeX is simply taking any input for Math rendering.
We split up into Text and Math parts. You can use a separator (eg. the common "$" or "$$").
Content between these separators is rendered as math while anything outside of these separators
is rendered as normal Flutter Text. This makes Text containing only some single formula parts much easier.

## Use this package as a library

Add this to your package's pubspec.yaml file:

```yaml
dependencies:
  katex_flutter: ^4.0.0+24
```
You can install packages from the command line with Flutter:

```shell
flutter pub get
```
Alternatively, your editor might support flutter pub get. Check the docs for your editor to learn more.

Now in your Dart code, you can use:

```dart
import 'package:katex_flutter/katex_flutter.dart';

...

// A static LaTeX block which may not change on `setState()`
return KaTeX(laTeXCode: Text("\\alpha", style: Theme.of(context)
                      .textTheme
                      .bodyText1
                      .copyWith(color: Colors.red)))

// A dynamic LaTeX block which is rebuilt on `setState()` (less efficient but required sometimes)
return Builder(builder: (context) => KaTeX(laTeXCode: Text("\\alpha", style: Theme.of(context)
                      .textTheme
                      .bodyText1
                      .copyWith(color: Colors.red))))
```

## Platform specific

***PLEASE NOTE: All the following changes are only required if you use our deprecated fallback library using
JavaScript. Otherwise you do NOT need to perform any of these changes.***

Please read [this document](DEPRECATED_JS_INTEGRATION.md) for information on our deprecated library.

## Source code

The source code is hosted on [GitLab](https://gitlab.com/testapp-system/katex_flutter). It's licensed under the terms and conditions of the [`EUPL-1.2`](LICENSE).

This package was initially created for the education project **[TestApp](https://testapp.schule/)**.


CHANGELOG.md
## 4.0.1+25 - Improved rendering of inline LaTeX

 * Moved from Wrap to RichText to provide better text rendering
 * Removed deprecated JS compatibility

## 4.0.0+24 - Moved to native Dart rendering instead of using JavaScipt

 * Moved all internal calls to the CaTeX library instead of JavaScipt
 * Increased rendering performance

## 3.3.0+23 - Stabilized Desktop support

 * Improved integration with `texlive` and `ImageMagick`

## 3.2.0+21 - Added custom font families

 * Inheriting text style from parent element / parenting build context

## 3.1.0+20 - Removed requirement for assets on Android and the web

 * You no longer need to copy the original KaTeX JS library into your app as it is now integrated into the package.

## 3.0.1+19 - Fixed Windows %TEMP% location

 * Fixed the temporary file's location on Windows

## 3.0.0+17 - Initial Desktop support

 * Support Fuchsia, macOS, Windows and Linux
 * Still requires local [TeX Live](https://www.tug.org/texlive/) and [ImageMagick](https://imagemagick.org/index.php) installation
 * Support for go-flutter

## 2.4.0+16 - Added offline support

 * Support for rendering equations offline on Android and the web
 * Support for custom delimiters on the web

## 2.3.0+14 - Implemented automatic width for web platform

 * Implemented automatic width for the web platform

## 2.2.0+12 - Fixed overflow on web platform

 * The compution ot the required width is now no longer limited to 200px (the initial width)

## 2.1.0+10 - Working dynamic boundaries on web platform

 * The web platform can now compute the Widget's width and height on it's own. No more need for `SizedBox` wrap.
 * Added dynamic generation of the required JavaScript code. Please update your `web/index.html` according to [README.md] and run `flutter pub run katex_flutter:main`

## 2.0.2+8 - Fixed equations displayed twice

 * Fixed some equations rendered twice due to inpropper MathML compatibility

## 2.0.1+7 - Fixed rendering of several equations

 * Fixed issues when using several different KaTeX-Widgets on web platform

## 2.0.0+6 - Support for web and mobile alongside

 * Separated mobile and web code
 * Updating the LaTeX code is not possible yet on mobile platforms
 * Automatic size and colors are not supported yet on web

## 1.1.1+5 - First support for web platforms

* Stable support for Android and iOS
* Experimental support for web platform. Auto-sizing the Widget is not supported yet on web.

## 0.0.1 - Initial release

* Not working yet. Just for testing.


lib/katex_flutter_mobile.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/widgets.dart';
import 'package:webview_flutter/webview_flutter.dart';

import 'katex_flutter.dart';

class KaTeXStateMobile extends State<KaTeX> {
  static const String _katexCDN = "https://cdn.jsdelivr.net/npm/katex/dist";
  String _lastKnownLaTeXCode = '';
  WebView _webView;

  // The controller is required to load content from a String
  WebViewController _controller;

  double _height = 50;
  double _width;

  @override
  void initState() {
    _lastKnownLaTeXCode = widget.laTeXCode.data;
    Set<JavascriptChannel> jsChannels = Set();
    jsChannels.add(JavascriptChannel(
        name: 'RenderedWebViewHeight',
        onMessageReceived: (JavascriptMessage message) {
          // Setting the widget's height to the height posted by JavaScript
          double viewHeight = double.parse(message.message) * 1.2;
          setState(() {
            _height = viewHeight;
          });
        }));
    if (!widget.inheritWidth)
      jsChannels.add(JavascriptChannel(
          name: 'RenderedWebViewWidth',
          onMessageReceived: (JavascriptMessage message) {
            // Setting the widget's width to the height posted by JavaScript
            double viewWidth = double.parse(message.message) * 1.3;
            setState(() {
              _width = viewWidth;
            });
          }));
    _webView = WebView(
      debuggingEnabled: false,
      javascriptMode: JavascriptMode.unrestricted,
      javascriptChannels: jsChannels,
      onWebViewCreated: (WebViewController webViewController) {
        _controller = webViewController;
        renderLaTeX();
      },
    );
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    if (widget.laTeXCode.data != _lastKnownLaTeXCode) renderLaTeX();
    return SizedBox(
      height: _height,
      width: _width,
      child: _webView,
    );
  }

  @override
  void dispose() {
    _webView = null;
    _controller = null;
    super.dispose();
  }

  String generateAppleHTMLCode({String laTeX}) {
    if (laTeX == null) laTeX = widget.laTeXCode.data;
    return '''<!DOCTYPE html>
<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover">
<link rel="stylesheet" href="$_katexCDN/katex.min.css">
<script defer src="$_katexCDN/katex.min.js"></script>
<script defer src="$_katexCDN/contrib/auto-render.min.js"></script>
<script>
  document.addEventListener("DOMContentLoaded", function() {
    renderMathInElement(document.querySelector("#katex_flutter"),{delimiters: [
      {left: "${widget.delimiter}", right: "${widget.delimiter}", display: false},
      {left: "${widget.displayDelimiter}", right: "${widget.displayDelimiter}", display: true}
    ]});
    var height = document.querySelector("#katex_flutter").clientHeight;
    var width = document.querySelector("#katex_flutter").clientWidth;
    if("RenderedWebViewHeight" in window) RenderedWebViewHeight.postMessage(height);
    if("RenderedWebViewWidth" in window) RenderedWebViewWidth.postMessage(width);
    });
</script>
<style>
:root {
  ${styleString(context, widget.laTeXCode)};
  background: #${widget.background.value.toRadixString(16).substring(2)}!important;
}
html, body {
  margin: 0;
  padding: 0;
}
body { overflow: auto; }
#katex_flutter {
  display: inline-block;
  width: auto;
  height: auto;
  overflow: auto;
}
</style>
</head>
<body><div id="katex_flutter">${widget.laTeXCode.data}</div></body>
</html>''';
  }

  void renderLaTeX() {
    _lastKnownLaTeXCode = widget.laTeXCode.data;
    String backgroundColorHex = '#' +
        widget.background.value
            .toRadixString(16)
            .substring(2)
            .replaceAll('+', '');
    if (Platform.isAndroid) {
      _controller.loadUrl(Uri(
          scheme: 'file',
          host: '',
          path:
              '/android_asset/flutter_assets/packages/katex_flutter/lib/katex_flutter.html',
          queryParameters: {
            'laTeX': widget.laTeXCode.data,
            'delimiter': widget.delimiter,
            'displayDelimiter': widget.displayDelimiter,
            'background': "$backgroundColorHex!important",
            'style': styleString(context, widget.laTeXCode),
          }).toString());
    } else {
      var localUri = Uri.dataFromString(generateAppleHTMLCode(),
          mimeType: 'text/html', encoding: utf8);
      _controller.loadUrl(localUri.toString());
    }
  }
}
